### ISSUE TITLE
AbstractStateNavigationRenderer#getRouteTarget does not work with CglibSubclassingInstantiationStrategy

### ISSUE THREAD

**Issue by TatuLund** [vaadin/flow#19100]
### Description of the bug

Due change in bean instantiation use `CglibSubclassingInstantiationStrategy` Spring Boot 3.2 (see `AbstractAutowireCapableBeanFactory`) `AbstractStateNavigationRenderer#getRouteTarget` will fail as `getClass` returns a proxy. Therefore `filter(component -> component.getClass().equals(routeTargetType))` does not match. 

### Expected behavior

The filter should match also with proxied classes when using Spring.

### Minimal reproducible example

-

### Versions

- Vaadin / Flow version: Vaadin 24.3

**Comment by TatuLund** [comment]
This is related to, but not the same as https://github.com/vaadin/flow/issues/16062

**Comment by Legioth** [comment]
Note that it was suggested in a separate discussion to change the `equals` check to use `isAssignableFrom` but this wouldn't be semantically correct in the hypothetical case of navigating between `class Foo extends Component` and `class Bar extends Foo` since it would then consider those two to be the same when going in one direction but different when going in the opposite direction.

**Comment by mcollovati** [comment]
We should perhaps inspect the component class hierarchy, search for the first not synthetic class, and then do the `equals` check against that one.

**Comment by TatuLund** [comment]
> We should perhaps inspect the component class hierarchy, search for the first not synthetic class, and then do the equals check against that one.

Exactly. In Spring there is utility for that https://github.com/spring-projects/spring-framework/blob/main/spring-core/src/main/java/org/springframework/util/ClassUtils.java#L987

But since AbstractStateNavigationRenderer should work also with projects without Spring or those using CDI, we can't use that directly. So we need our own utility as you described.

**Comment by mcollovati** [comment]
This is what I used as a replacement for Spring `ClassUtils.getUserClass()` in Quarkus.

https://github.com/mcollovati/quarkus-hilla/blob/0f1dba746235de456bdee47bea69df9c19cec3ae/runtime-commons/src/main/java/com/github/mcollovati/quarkus/hilla/SpringReplacements.java#L30-L38

**Comment by mcollovati** [comment]
The issue was assigned to a developer and is currently being investigated

**Comment by mcollovati** [comment]
`isSyntethc` works with Quarkus generated proxy, but not with Spring using CGLIB. Nor it will with ByteBuddy generated subclasses.
So, in addition to the `isSynthetic` check, we should probably do additional verification on class name, as for example the mentioned Spring `ClassUtils.getUserClass()` method does.

**Comment by mcollovati** [comment]
I can see two potential solutions:

1. define a method in `AbstractNavigationStateRenderer` (or `ReflectTools`) that tries to detect the base class in case of proxies, applying some hard-coded attempts based on class name (e.g. contains `$$` like in `ClassUtils.getUserClass()` or `$ByteBuddy$`, etc) and `Class.isSynthetic()` flag.
2. Add a default `Class<?> getUserClass(Class<?> clazz)` method to `Instantiator` and have a specific implementation based on the expected execution environment. The default implementation should just return the input class, `SpringInstantiator` can delegate to `ClassUtils.getUserClass()`, CDI and Quarkus should probably work checking `isSynthetic`.

Given that, the comparison we do in `AbstractNavigationStateRenderer` should be against objects created by `Instantiator`, I would go probably with option 2, but only if such a change can be back-ported to 24.3.

**Comment by mcollovati** [comment]
@TatuLund could you provide an example use case that shows how the current behavior can break an application? I'm asking because I'd like to add some sensible integration tests in Flow.

**Comment by Legioth** [comment]
Seems to me like it would make sense to add a method to `Instantiator`. The `isSynthetic` approach is probably generic enough that it could be used in the default implementation so that CDI and Quarkus implementations wouldn't even need any additional changes.
